pkg/scheduler/internal/queue/scheduling_queue.go

```diff
@@ -916,3 +918,103 @@ func MakeNextPodFunc(queue SchedulingQueue) func() *v1.Pod {
 func podInfoKeyFunc(obj interface{}) (string, error) {
 	return cache.MetaNamespaceKeyFunc(obj.(*podInfo).pod)
 }
+
+// PodBackoffMap is a structure that stores backoff related information for pods
+type PodBackoffMap struct {
+	// lock for performing actions on this PodBackoffMap
+	lock sync.Mutex
+	// initial backoff duration
+	initialDuration time.Duration
+	// maximal backoff duration
+	maxDuration time.Duration
+	// map for pod -> number of attempts for this pod
+	podAttempts map[ktypes.NamespacedName]int
+	// map for pod -> lastUpdateTime pod of this pod
+	podLastUpdateTime map[ktypes.NamespacedName]time.Time
+}
+
+// NewPodBackoffMapWithDefaultTime creates a default PodBackoffMap.
+func NewPodBackoffMapWithDefaultTime() *PodBackoffMap {
+	return NewPodBackoffMap(1*time.Second, 60*time.Second)
+}
+
+// NewPodBackoffMap creates a PodBackoffMap with initial duration and max duration.
+func NewPodBackoffMap(initialDuration, maxDuration time.Duration) *PodBackoffMap {
+	return &PodBackoffMap{
+		initialDuration:   initialDuration,
+		maxDuration:       maxDuration,
+		podAttempts:       make(map[ktypes.NamespacedName]int),
+		podLastUpdateTime: make(map[ktypes.NamespacedName]time.Time),
+	}
+}
+
+// GetBackoffTime returns the time that nsPod completes backoff
+func (pbm *PodBackoffMap) GetBackoffTime(nsPod ktypes.NamespacedName) (time.Time, bool) {
+	pbm.lock.Lock()
+	defer pbm.lock.Unlock()
+	if _, found := pbm.podAttempts[nsPod]; found == false {
+		return time.Time{}, false
+	}
+	lastUpdateTime := pbm.podLastUpdateTime[nsPod]
+	backoffDuration := pbm.calculateBackoffDuration(nsPod)
+	backoffTime := lastUpdateTime.Add(backoffDuration)
+	return backoffTime, true
+}
+
+// TryBackoffAndWait tries to perform backoff for a non-preempting pod.
+// it is invoked from factory.go if util.PodPriorityEnabled() returns false.
+func (pbm *PodBackoffMap) TryBackoffAndWait(nsPod ktypes.NamespacedName, stop <-chan struct{}) bool {
+	pbm.lock.Lock()
+	defer pbm.lock.Unlock()
+	backoffDuration := pbm.calculateBackoffDuration(nsPod)
+	select {
+	case <-time.After(backoffDuration):
+		return true
+	case <-stop:
+		return false
+	}
+}
+
+// calculateBackoffDuration is a helper function for calculating the backoffDuration
+// based on the number of attempts the pod has made.
+func (pbm *PodBackoffMap) calculateBackoffDuration(nsPod ktypes.NamespacedName) time.Duration {
+	backoffDuration := pbm.initialDuration
+	if _, found := pbm.podAttempts[nsPod]; found {
+		for i := 1; i < pbm.podAttempts[nsPod]; i++ {
+			backoffDuration = backoffDuration << 1
+			if backoffDuration > pbm.maxDuration {
+				return pbm.maxDuration
+			}
+		}
+	}
+	return backoffDuration
+}
+
+// ClearPodBackoff removes all tracking information for nsPod.
+// Lock is supposed to be acquired by caller.
+func (pbm *PodBackoffMap) clearPodBackoff(nsPod ktypes.NamespacedName) {
```

Comment:
How about just keep one clearPodBackoff()?

This method seems to be just a variant of the method in L382.

Reply:
It's fine to keep it.